### PR TITLE
adding reconcile request id to better track multiple thread logs

### PR DESCRIPTION
**Pull request by RamLavi**
Signed-off-by: Ram Lavi <ralavi@redhat.com>

**What this PR does / why we need it**:
Currently it's hard to follow along the multiple reconcile logs, that sometimes reconcile the same vm (for example, a create followed by a update request).
In order to make it easier to follow the logs, we added a pseudo random "unique" ID, to be attached to the start of all relevant logs in reconcile. 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/virtualmachine/virtualmachine_controller.go [hunk 1]
```diff
@@ -108,19 +111,19 @@ func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result
...
 	logger.V(1).Info("The VM is being marked for deletion")
-	err = r.removeFinalizerAndReleaseMac(instance, &request)
+	err = r.removeFinalizerAndReleaseMac(instance, &request, reconcileRequestId)
 	return reconcile.Result{}, err
 }
 
-func (r *ReconcilePolicy) addFinalizerAndUpdate(request *reconcile.Request) error {
-	logger := log.WithName("addFinalizerAndUpdate").WithValues("virtualMachineName", request.Name, "virtualMachineNamespace", request.Namespace)
+func (r *ReconcilePolicy) addFinalizerAndUpdate(request *reconcile.Request, RequestId int) error {
+	logger := log.WithName("addFinalizerAndUpdate").WithValues("RequestId", RequestId, "virtualMachineName", request.Name, "virtualMachineNamespace", request.Namespace)
```
Comment (phoracek) [inline pkg/controller/virtualmachine/virtualmachine_controller.go:126]: Would it make sense to pass the whole logger and just extend it with the info about finalizer? It is still the same transaction. That way finalizer does not need to know about any transaction I'd being a thing. The same for other instances where you pass the id.

Passing the whole logger wuod also make logs clearer. You would see that this finalizer function has been called as a part of the reconciliation.

Comment (RamLavi) [reply pkg/controller/virtualmachine/virtualmachine_controller.go:126]: you mean that the same logger instance is passed from the reconcile function, and each time I concat the `log.WithName(<func name>).WithValues(bla bla..)` to it?

Comment (phoracek) [reply pkg/controller/virtualmachine/virtualmachine_controller.go:126]: I would pass the logger from Reconcile to addFinalizerAndUpdate as an argument. I'm not sure you can add another WithName. Maybe just WithValues("operation", "adding finalizer") or "function", "addFinalizerAndUpdate"

Comment (RamLavi) [reply pkg/controller/virtualmachine/virtualmachine_controller.go:126]: what do you think about adding it with `WithName`? that way in the log name I get the entire stack path: `VirtualMachine Controller.Reconcile.addFinalizerAndUpdate...`

**Comment by qinqon** [comment]
@RamLavi can you paste here an example of how the log looks like ? I want to see how multiple `WithName` works

**Comment by RamLavi** [comment]
> @RamLavi can you paste here an example of how the log looks like ? I want to see how multiple `WithName` works

```
2020-06-01T07:08:30.784Z	DEBUG	VirtualMachine Controller.Reconcile	got a virtual machine event in the controller	{"RequestId": 4751997750760398084, "virtualMachineName": "testvmbmjxhqm2lnsqd98q7mz8g49tqskvjvjt", "virtualMachineNamespace": "kubemacpool-test"}
2020-06-01T07:08:30.789Z	DEBUG	VirtualMachine Controller.Reconcile	The VM is not being deleted	{"RequestId": 4751997750760398084, "virtualMachineName": "testvmbmjxhqm2lnsqd98q7mz8g49tqskvjvjt", "virtualMachineNamespace": "kubemacpool-test"}
2020-06-01T07:08:30.789Z	DEBUG	VirtualMachine Controller.Reconcile.addFinalizerAndUpdate	The VM does not have a finalizer	{"RequestId": 4751997750760398084, "virtualMachineName": "testvmbmjxhqm2lnsqd98q7mz8g49tqskvjvjt", "virtualMachineNamespace": "kubemacpool-test"}
2020-06-01T07:08:30.791Z	DEBUG	controller-runtime.webhook.webhooks	received request	{"webhook": "/mutate-virtualmachines", "UID": "56c0b313-ae49-4679-812f-9855c802c3d8", "kind": "kubevirt.io/v1alpha3, Kind=VirtualMachine", "resource": {"group":"kubevirt.io","version":"v1alpha3","resource":"virtualmachines"}}

```

File: pkg/pool-manager/virtualmachine_pool.go [hunk 2]
```diff
@@ -98,35 +99,32 @@ func (p *PoolManager) AllocateVirtualMachineMac(virtualMachine *kubevirt.Virtual
 	return nil
 }
 
-func (p *PoolManager) ReleaseVirtualMachineMac(vm *kubevirt.VirtualMachine) error {
+func (p *PoolManager) ReleaseVirtualMachineMac(vm *kubevirt.VirtualMachine, reconcileLog logr.Logger) error {
```
Comment (qinqon) [inline pkg/pool-manager/virtualmachine_pool.go:102]: Hey what do you think about removing log from here and creating the following at PoolManager
```golang
func (p *PoolManager) String() string {
...
}
```

This way you can do something like the following

```golang
fmt.Sprintf("KMP pool manager: %s", poolManager)
```

File: pkg/controller/virtualmachine/virtualmachine_controller.go [hunk 3]
```diff
@@ -108,19 +112,19 @@ func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result
 		// The object is not being deleted, so if it does not have a finalizer,
 		// then lets add the finalizer and update the object.
 		if !helper.ContainsString(instance.ObjectMeta.Finalizers, pool_manager.RuntimeObjectFinalizerName) {
-			err = r.addFinalizerAndUpdate(&request)
+			err = r.addFinalizerAndUpdate(&request, logger)
```
Comment (qinqon) [inline pkg/controller/virtualmachine/virtualmachine_controller.go:115]: I am tented to sugget to have requestLogger as member of `PolicyReconcile` since we have only one goroutine running `Reconcile` and finalizer functions, but that's wrong if we do so and we change that we are broken, the alternative so we don't pass the logger to functions would be to add a struct with `PolicyReconcile`and logger and addFinalizer and removeFinalizers are functions from that struct.

File: pkg/pool-manager/virtualmachine_pool.go (hunk 2 above)
Comment (phoracek) [reply pkg/pool-manager/virtualmachine_pool.go:102]: How does it help with the identification of single execution?

File: pkg/controller/virtualmachine/virtualmachine_controller.go (hunk 3 above)
Comment (phoracek) [reply pkg/controller/virtualmachine/virtualmachine_controller.go:115]: The point of this PR is to add unique ID per each execution of Reconcile, so we can track those.

File: pkg/pool-manager/virtualmachine_pool.go (hunk 2 above)
Comment (qinqon) [reply pkg/pool-manager/virtualmachine_pool.go:102]: Not much, just passing the logger to all functions as an specific case of `sillver bullet` feels uncomplete, maybe we need to pass some kind of requestContext to all the calls here so we can really trace things.

Comment (phoracek) [reply pkg/pool-manager/virtualmachine_pool.go:102]: I hate the appeal of dependency injection (countless arguments passed down the stack everywhere), but what is better context than single purpose obejct (logger) doing one thing, keeping all the information while allowing its users to make transparent logs without understanding that they are part of Reconcile loop

File: pkg/controller/virtualmachine/virtualmachine_controller.go [hunk 4]
```diff
@@ -108,19 +112,19 @@ func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result
...
 	// The object is being deleted
 	logger.V(1).Info("The VM is being marked for deletion")
-	err = r.removeFinalizerAndReleaseMac(instance, &request)
+	err = r.removeFinalizerAndReleaseMac(instance, &request, logger)
 	return reconcile.Result{}, err
 }
 
-func (r *ReconcilePolicy) addFinalizerAndUpdate(request *reconcile.Request) error {
-	logger := log.WithName("addFinalizerAndUpdate").WithValues("virtualMachineName", request.Name, "virtualMachineNamespace", request.Namespace)
+func (r *ReconcilePolicy) addFinalizerAndUpdate(request *reconcile.Request, reconcileLog logr.Logger) error {
```
Comment (phoracek) [inline pkg/controller/virtualmachine/virtualmachine_controller.go:126]: It does not matter to this method that the logger is from reconciler. For us it is just a logger. Maybe `parentLogger`?

File: pkg/controller/virtualmachine/virtualmachine_controller.go [hunk 5]
```diff
@@ -143,18 +147,18 @@ func (r *ReconcilePolicy) addFinalizerAndUpdate(request *reconcile.Request) erro
 
 	logger.V(1).Info("Finalizer was added to the VM")
 
-	return r.poolManager.MarkVMAsReady(virtualMachine)
+	return r.poolManager.MarkVMAsReady(virtualMachine, reconcileLog)
 }
 
-func (r *ReconcilePolicy) removeFinalizerAndReleaseMac(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request) error {
-	logger := log.WithName("removeFinalizerAndReleaseMac").WithValues("virtualMachineName", request.Name, "virtualMachineNamespace", request.Namespace)
+func (r *ReconcilePolicy) removeFinalizerAndReleaseMac(virtualMachine *kubevirt.VirtualMachine, request *reconcile.Request, reconcileLog logr.Logger) error {
```
Comment (phoracek) [inline pkg/controller/virtualmachine/virtualmachine_controller.go:153]: ditto

File: pkg/pool-manager/virtualmachine_pool.go (hunk 2 above)
Comment (phoracek) [inline pkg/pool-manager/virtualmachine_pool.go:102]: ditto

File: pkg/pool-manager/virtualmachine_pool.go [hunk 6]
```diff
@@ -98,35 +99,32 @@ func (p *PoolManager) AllocateVirtualMachineMac(virtualMachine *kubevirt.Virtual
...
 	p.poolMutex.Lock()
 	defer p.poolMutex.Unlock()
-	log.V(1).Info("ReleaseVirtualMachineMac: data",
+	logger.V(1).Info("data",
 		"macmap", p.macPoolMap,
 		"podmap", p.podToMacPoolMap,
 		"currentMac", p.currentMac.String())
 
 	if len(vm.Spec.Template.Spec.Domain.Devices.Interfaces) == 0 {
-		log.V(1).Info("no interfaces found for virtual machine, skipping mac release",
```
Comment (phoracek) [inline pkg/pool-manager/virtualmachine_pool.go:110]: Here the passed logger starts to be a little weird. We assume that it contains VM info, so we are dropping it from here. But there is no guarantee somebody removes it from the parent logger. I'm puzzled. I don't think we can do anything about it. I don't like the fact that it semantically glues the whole chain together :grimacing: Any suggestion?

File: pkg/pool-manager/virtualmachine_pool.go [hunk 7]
```diff
@@ -417,8 +415,8 @@ func (p *PoolManager) AddMacToWaitingConfig(allocations map[string]string) error
 }
 
 // Remove all the mac addresses from the waiting configmap this mean the vm was saved in the etcd and pass validations
-func (p *PoolManager) MarkVMAsReady(vm *kubevirt.VirtualMachine) error {
-	logger := log.WithName("MarkVMAsReady").WithValues("macPoolMap", p.macPoolMap, "virtualMachineNamespace", vm.Namespace, "virtualMachineName", vm.Name)
+func (p *PoolManager) MarkVMAsReady(vm *kubevirt.VirtualMachine, reconcileLog logr.Logger) error {
```
Comment (phoracek) [inline pkg/pool-manager/virtualmachine_pool.go:418]: ditto

File: pkg/pool-manager/virtualmachine_pool.go (hunk 2 above)
Comment (qinqon) [reply pkg/pool-manager/virtualmachine_pool.go:102]: ack, fair enough.

File: pkg/pool-manager/virtualmachine_pool.go (hunk 6 above)
Comment (RamLavi) [reply pkg/pool-manager/virtualmachine_pool.go:110]: well for better and for worse, the reason we did this is because there is information (the unique id, the vm info) that we hold in the first reconcile function, that we want to remember while reading the logs but don't necessarily want to carry with us to each future function. 
But I think that since the both (the vm info and the id) are consistent and always relevant through out the Reconcile function - then it's ok. what do you think?

**Comment by qinqon** [comment]
@RamLavi @phoracek maybe you kill me but the golang `context` package is also designed to pass request related info across different APIs [1], maybe we can pass there whatever we need like request ID and the like, so we can log it and it's transparent.

[1] https://golang.org/pkg/context/#pkg-variables

**Comment by phoracek** [comment]
@qinqon I don't want to carry context ID all the way down to helper functions. Helpers should not know or care about that value. They just want to log, so let's give them a logger. Sharing of context with this value sounds like unnecessary coupling.

File: pkg/pool-manager/virtualmachine_pool.go (hunk 6 above)
Comment (phoracek) [reply pkg/pool-manager/virtualmachine_pool.go:110]: I hate that we depend on implicitly given context, but I don't have a better option to propose. So I guess it's ok

**Comment by phoracek** [comment]
Seems to be already backported.
